src/trex/quality_control.py

```diff
@@ -0,0 +1,375 @@
+        return df.groupby(['#cell_id']).umi.agg('count')
+    else:
+        return df.groupby(['cell_id']).counts.sum()
+
+
+def get_molecules_per_clone_ids(df: pd.DataFrame,
+                               molecules_dataframe: bool = True) -> pd.Series:
+    """Get a pandas Series with the number of CloneID molecules found per unique
+    CloneID. molecules_dataframe is set to True by default, if a cells
+    DataFrame is being used, then this must be False."""
+    if molecules_dataframe:
+        return df.groupby(['clone_id']).umi.agg('count')
+    else:
+        return df.groupby(['clone_id']).counts.sum()
+
+
+def get_unique_clone_ids_per_cell(df: pd.DataFrame,
+                                 molecules_dataframe: bool = True) -> pd.Series:
+    """Get a pandas Series with the number of unique CloneID molecules found per
+    cell. molecules_dataframe is set to True by default, if a cells DataFrame is
+    being used, then this must be False."""
+    if molecules_dataframe:
+        return df.groupby('#cell_id').clone_id.unique().apply(len)
+    else:
+        return df.groupby(['cell_id']).clone_id.unique().apply(len)
+
+
+def add_clone_ids_per_clone(clones: pd.DataFrame,
+                           cells_filtered: pd.DataFrame) -> pd.DataFrame:
+    """Add  the unique CloneID molecules found in each clone to the clones
+    DataFrame. clones is the clones dataframe from clones.txt and cells_filtered
+    is the dataframe containing CloneIDs per cell (cells_filtered.txt)."""
+    cells_clone_id = cells_filtered.groupby('cell_id').clone_id.apply(set)
+    clones['clone_ids'] = clones.cell_id.apply(lambda x: cells_clone_id[x])
+    return clones
+
+
+def get_clone_ids_per_clone(clones: pd.DataFrame) -> pd.Series:
+    """Get a pandas Series with the number of unique CloneIDs found per
+    clone."""
+    return clones.groupby('#clone_id').clone_ids.apply(
+        lambda x: len(set.union(*x)))
+
+
+def get_clone_sizes(clones: pd.DataFrame) -> pd.Series:
+    """Get a pandas Series with the umber of cells found in each
+    clone. clones is the clones dataframe from clones.txt."""
+    return clones.groupby('#clone_id').cell_id.count()
+
+
+def plot_discrete_histogram(series: pd.Series,
+                            title: str = None,
+                            xlabel: str = None, ylabel: str = None,
+                            txt:str = None,
+                            ax: plt.Axes = None):
+    """plots a discrete histogram with optional title, x and y labels, and added
+     text below."""
+    if ax is None:
+        fig, ax = plt.subplots(1, 1)
+
+    ax = sns.histplot(series, discrete=True, log=True, ax=ax)
+
+    ax.set_xlabel(xlabel)
+    ax.set_ylabel(ylabel)
+    ax.set_title(title)
+
+    if txt is not None:
+        plt.sca(ax)
+        plt.text(0, -0.2, txt, transform=ax.transAxes, size=12, wrap=True,
+                 ha='left', va='top')
+
+    return ax
+
+
+def length_read(molecules: pd.DataFrame,
```

Comment:
It’s unclear from the function names why there’s both a `get_length_read` and a `length_read` function. This one should probably have `plot` in its name.